source/symbols/Type.cpp

```diff
@@ -239,6 +239,13 @@ bool Type::isUnpackedArray() const {
     }
 }
 
+bool Type::isUnpackedArrayOfByte() const {
```

Comment:
The Type::isByteArray method already does much the same thing, though it only allows actual byte arrays but does allow dynamic arrays. The LRM isn't very clear on this point, but I tend to think it should only allow the byte type, but should allow dynamic arrays and queue as well. Thoughts?

Reply:
I plan to remove isUnpackedArrayOfByte and replace it with
```c++
type.isByteArray() && type.getCanonicalType().hasFixedRange()
```
To be safe, I will limit the enhancement to fixed size packed array for now. Only using isByteArray will accept SystemVerilog code like
```sv
localparam byte d[] = "hi2";
```
and it is a question how to determine at compile time the size of a dynamic array whose size can be changed at runtime. Please let me know if you think it is better to have a new function than to use existing isByteArray() && hasFixedRange().

Reply:
Hmm, I think we should either make this new code support dynamic arrays as well, or we should remove the dynamic array support from isByteArray(). It's ok for a parameter to generally be a dynamic array, the code can already resolve it at compile time based on the assignment, so it should be possible to make it work for string literals as well (it should be sized according to the number of characters in the string). The question is whether SystemVerilog allows that according to the LRM, and I think it does because it uses the term "unpacked array" to mean all forms of unpacked arrays (fixed size, dynamic, and queues). 

I do think isByteArray() needs to be changed to disallow associative arrays, because those clearly don't work. You could make that change or I can do it.

I think you don't have to do much extra work to support dynamic arrays, it should really just work as-is. For queues, your convert function will need to check whether to.isQueue() and create an SVQueue instead of Elements vector.

Reply:
What you said makes sense. My solution is

1. In isByteArray(), reject associative arrays by adding
```c++
    if (ct.kind == SymbolKind::AssociativeArrayType)
        return false;
```
2. Convert string to byte queue by adding  ConstantValue::convertToByteQueue(bool isSigned) const;
```c++
ConstantValue ConstantValue::convertToByteQueue(bool isSigned) const {
    if (isQueue())
        return *this;
    std::string result;
    if (isInteger())
        result = convertToStr().str();
    else if (isString())
        result = str();
    else
        return nullptr;
    SVQueue queue;
    for (auto ch : result)
        queue.emplace_back(SVInt(8, static_cast<uint64_t>(ch), isSigned));
    return queue;
}
```
3. In ConversionExpression::convert, pass size=0 to convertToByteArray for dynamic arrays, and call convertToByteQueue for queues.
```c++
    if (to.isByteArray()) {
        const auto& ct = to.getCanonicalType();
        const auto isSigned = ct.getArrayElementType()->isSigned();
        if (ct.isQueue()) {
            return value.convertToByteQueue(isSigned);
        }
        else {
            bitwidth_t size;
            if (ct.hasFixedRange())
                size = ct.as<FixedSizeUnpackedArrayType>().range.width();
            else
                size = 0;  // dynamic array use string size
            return value.convertToByteArray(size, isSigned);
        }
    }
```
4. Inside convertToByteArray, if size==0 from dynamic arrays, set size to string size:
```c++
ConstantValue ConstantValue::convertToByteArray(bitwidth_t size, bool isSigned) const {
    if (isUnpacked())
        return *this;
    std::string result;
    if (isInteger())
        result = convertToStr().str();
    else if (isString())
        result = str();
    else
        return nullptr;
    Elements array;
    if (!size)  // dynamic array use string size
        size = static_cast<bitwidth_t>(result.size());
    array.reserve(size);
    for (auto ch : result) {
        if (array.size() >= size)
            break;
        array.emplace_back(SVInt(8, static_cast<uint64_t>(ch), isSigned));
    }
    while (array.size() < size)
        array.emplace_back(SVInt(8, 0, isSigned));
    return array;
}
```
5. With these, I can use type.canBeStringLike() to replace type.isString() to accept such assignments.

A SV test case I created to test these C++ code is
```sv
module test;
  localparam byte d[] = "dynamic array";
  localparam byte q[$] = "queue";
  localparam dynamicSize = d.size;
  localparam queueSize = q.size;
  // Asscoiative array should fail
  // localparam byte a[string] = "associative";
endmodule
```

Reply:
Yeah, this all sounds good to me. Thanks!